hbase-server/src/main/java/org/apache/hadoop/hbase/master/MasterFileSystem.java

```diff
@@ -355,9 +354,9 @@ private void checkStagingDir() throws IOException {
         if (!this.fs.mkdirs(p, HiddenDirPerms)) {
           throw new IOException("Failed to create staging directory " + p.toString());
         }
-      } else {
-        this.fs.setPermission(p, HiddenDirPerms);
       }
+      this.fs.setPermission(p, HiddenDirPerms);
```

Comment:
In this case, `staging` dir would have already been created with `HiddenDirPerms` perf i.e. `this.fs.mkdirs(p, HiddenDirPerms)` should be true for this statement to be executed. So, isn't this statement redundant? We already created staging dir with HiddenDirPerms permission and now we are again setting `HiddenDirPerms` to `staging` dir.

Reply:
It seems that staging does use HiddenDirPerms for creation, however this permission will be affected by umask in DFSClient. If fs is DistributedFileSystem, if we set too strict umask, such as 777, the final created folder will not have any permissions. I think there might be some API misuse here, in mkdirs(FileSystem fs, Path dir, FsPermission permission) and mkdirs(Path f, FsPermission permission).[mkdirs(FileSystem fs, Path dir, FsPermission permission)](https://github.com/apache/hadoop/blob/d346be9bd2d45d0b17ecf0c2b9d8466f00b47e0d/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/FileSystem.java#L746) and [mkdirs(Path f, FsPermission permission)](https://github.com/apache/hadoop/blob/d346be9bd2d45d0b17ecf0c2b9d8466f00b47e0d/hadoop-hdfs-project/hadoop-hdfs-client/src/main/java/org/apache/hadoop/hdfs/DistributedFileSystem.java#L1471)

Reply:
And I found a similar situation inside DistributedFileSystem, setPermission after mkdir[provisionEZTrash](https://github.com/apache/hadoop/blob/d346be9bd2d45d0b17ecf0c2b9d8466f00b47e0d/hadoop-hdfs-project/hadoop-hdfs-client/src/main/java/org/apache/hadoop/hdfs/DistributedFileSystem.java#L2971).I think we should make sure that the file is created with the correct expected permissions, in order to ensure the normal operation of the project, it is necessary for us to confirm its permissions once the creation is complete.

Reply:
Makes sense, thanks for the detailed explanation.